esp32: improve toolchain management

sylvioalves opened the pull request.

sylvioalves committed:
toolchain: esp32: define target toolchain

In order to retrieve TOOLCHAIN_NAME from Kconfig,
generic.cmake points to any toolchain as a
boilerplate implementation. Now, target.cmake file handles
proper toolchain selection.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

sylvioalves committed:
docs: esp32: update environment setup content

Add toolchain path updates into documentation.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

sylvioalves commented:
Rebased to fix conflict. PR ready for review.

cfriedt commented:
@sylvioalves - I'm ok with this going in as there are no code changes and it is mainly toolchain-related. Have you run `twister`  successfully with some representative hardware in the loop? I would like to know with some certainty that this change will not introduce any regressions.

cfriedt reviewed (approved):
This is fine with me, as long as @sylvioalves  can run `twister` to ensure we have test coverage on areas not covered by CI.

sylvioalves commented:
> @sylvioalves - I'm ok with this going in as there are no code changes and it is mainly toolchain-related. Have you run `twister` successfully with some representative hardware in the loop? I would like to know with some certainty that this change will not introduce any regressions not covered by CI.

@cfriedt As far as I know, there is not CI for Espressif devices regarding hardware tests (yet).  Zephyr's CI also does not run twister for variant `espressif`.

cfriedt commented:
> @cfriedt As far as I know, there is not CI for Espressif devices regarding hardware tests (yet). Zephyr's CI also does not run twister for variant `espressif`.

Have you been able to test this out with fairly decent coverage? No compiler / linker bugs or things of that sort?

sylvioalves commented:
> > @cfriedt As far as I know, there is not CI for Espressif devices regarding hardware tests (yet). Zephyr's CI also does not run twister for variant `espressif`.
> 
> @sylvioalves Have you been able to test this out on hardware with fairly decent coverage? No compiler / linker bugs or things of that sort?

Yes. It is very straightforward. Let me bring up some background:  
Currently, ESP32 toolchain path needs to be exported for each SOC as below:  
`export ESPRESSIF_TOOLCHAIN_PATH="${HOME}/.espressif/tools/riscv32-esp-elf/1.24.0.123_64eb9ff-8.4.0/riscv32-esp-elf"`
`export ESPRESSIF_TOOLCHAIN_PATH="${HOME}/.espressif/tools/xtensa-esp32-elf/esp-2020r3-8.4.0/xtensa-esp32-elf"`

This is difficult to maintain for new SoCs and toolchain update. So this PR makes a common toolchain path, which enables build projects for esp32, esp32s2, esp32c3 (and so) without having to change `ESPRESSIF_TOOLCHAIN_PATH` at all.  
`export ESPRESSIF_TOOLCHAIN_PATH="${HOME}/.espressif/tools/zephyr"`

cfriedt commented:
cc @tejlmand @gmarull - this mainly touches docs & build. No code paths, but I see it as a significant improvement that should probably be in v2.7.0. Are you able to ack?

cfriedt commented:
@nashif - should this wait for TSC approval, or are we able to merge before then?

tejlmand reviewed (changes requested):
to avoid this being merged.
Will post comments soon.

tejlmand commented on soc/riscv/esp32c3/Kconfig.defconfig:
```diff
@@ -8,6 +8,10 @@ if SOC_ESP32C3
 
+config SOC_TOOLCHAIN_NAME
+	string
+	default "riscv32-esp-elf"
```
You know the arch and you know the soc, so it should be possible to deduce this in similar way as is done for the Zephyr SD here: https://github.com/zephyrproject-rtos/sdk-ng/blob/main/cmake/zephyr/target.cmake

sylvioalves replied:
Sure, I'll take it.

tejlmand commented on cmake/toolchain/espressif/target.cmake:
```diff
@@ -1,3 +1,21 @@
+set(LINKER ld)
+set(BINTOOLS gnu)
+
+set(CROSS_COMPILE_TARGET ${SOC_TOOLCHAIN_NAME})
```
A Kconfig setting must be prefixed with `CONFIG_` when being used in CMake, but why not do like this:
```suggestion
set(CROSS_COMPILE_TARGET_xtensa_esp32    xtensa-esp32-elf)
set(CROSS_COMPILE_TARGET_xtensa_esp32s2  xtensa-esp32s2-elf)
set(CROSS_COMPILE_TARGET_riscv_esp32c3   riscv32-esp-elf)

set(CROSS_COMPILE_TARGET ${CROSS_COMPILE_TARGET_${ARCH}_${CONFIG_SOC}})
set(SYSROOT_TARGET       ${CROSS_COMPILE_TARGET})
```

That will keep the relation between the toolchain and the soc in same place as is done for the Zephyr SDK, and not start to move such knowledge into Kconfig.

sylvioalves replied:
@tejlmand thanks for the input, your suggestion is just pretty much better.

tejlmand replied:
Thanks for the update.

Code wise i'm ready for +1, just a small observation I would like some feedback on.

tejlmand reviewed (changes requested):
small proposal that will make it possible to drop: https://github.com/zephyrproject-rtos/zephyr/pull/38540/commits/a625c272a88f450e02a36003d6b5e60573fa362d

tejlmand commented on boards/riscv/esp32c3_devkitm/doc/index.rst:
```diff
@@ -51,21 +51,19 @@ On Linux and macOS:
 
    export ZEPHYR_TOOLCHAIN_VARIANT="espressif"
-   export ESPRESSIF_TOOLCHAIN_PATH="${HOME}/.espressif/tools/riscv32-esp-elf/1.24.0.123_64eb9ff-8.4.0/riscv32-esp-elf"
+   export ESPRESSIF_TOOLCHAIN_PATH="${HOME}/.espressif/tools/zephyr"
```
just curious.

It seems the espressif toolchain is now expected to be installed under a different folder per default.
It that part of the espressif installer itself, or is there some missing doc update on toolchain installation ?

Also what about users having already having the espressif toolchain installed ?
How will they be aware that they must now move their current
`.espressif/tools/riscv32-esp-elf` into `.espressif/tools/zephyr/riscv32-esp-elf` instead.

Also, is `1.24.0.123_64eb9ff-8.4.0` no longer part of the toolchain installation path ?

sylvioalves replied:
@tejlmand I have explained that already in this PR. Can you check previous comments? The main reason is having the same export PATH for Espressif toolchain to all SoCs, not depending on folder names and versioning as it is right now.

So, as you can see in the document update in this PR, no more `1.24.0.123_64eb9ff-8.4.0` or toolchain name explicit in folder path.  

`west espressif install` is the download agent responsible to download and extract the toolchain in proper place (or user can manually download it as well).  

> Also what about users already having the espressif toolchain installed ?
When building, it will give them cmake error with a message to re-configure toolchain path. 
Is there any recommendation on how to inform users about this change besides changing the ESP32 documentation?

sylvioalves replied:
This is the output when running  `west espressif install`. It also informs the user how to export path into the environment.
```
sylvio@pc:~/projects/espressif/zephyr/zephyr$ west espressif install
=== downloading ESP-IDF tools..
Installing tools: xtensa-esp32-elf, xtensa-esp32s2-elf, riscv32-esp-elf, openocd-esp32
Skipping xtensa-esp32-elf@esp-2020r3-8.4.0 (already installed)
Skipping xtensa-esp32s2-elf@esp-2020r3-8.4.0 (already installed)
Skipping riscv32-esp-elf@1.24.0.123_64eb9ff-8.4.0 (already installed)
Skipping openocd-esp32@v0.10.0-esp32-20200709 (already installed)
=== downloading ESP-IDF tools completed
The toolchain has been downloaded to /home/sylvio/.espressif/tools/zephyr
Export or set the following variables into the environment:
export ESPRESSIF_TOOLCHAIN_PATH="/home/sylvio/.espressif/tools/zephyr"
export ZEPHYR_TOOLCHAIN_VARIANT="espressif"
```

tejlmand replied:
> @tejlmand I have explained that already in this PR.

If you refer to this comment https://github.com/zephyrproject-rtos/zephyr/pull/38540#issuecomment-922276759, then you only describe why, but my question was about who / how that path was ensured.
The installer or a guide ?

Now that question you answer with the `west espressif install` sentence. 

> > Also what about users already having the espressif toolchain installed ?
>
> When building, it will give them cmake error with a message to re-configure toolchain path.

No, you give them an error message requesting users to re-install something that is already installed.
I think we should ensure that if users have a setup with the old style then that will still be supported.

That will allow users to keep working and update the toolchain when a new revision is released and they update anyway.

By the way, your `west espressif install` fails on my machine.
```
west espressif install
=== downloading ESP-IDF tools..
/usr/bin/env: ‘python’: No such file or directory
FATAL ERROR: command exited with status 127: ./tools/idf_tools.py --tools-json=tools/zephyr_tools.json install
```
because I only have `python3` and no `python` link.
Why not use the python that was already invoked with `west` ?

sylvioalves commented on cmake/toolchain/espressif/generic.cmake:
```diff
@@ -12,20 +12,29 @@ set(BINTOOLS gnu)
+if(NOT CROSS_COMPILE_TARGET)
+  message(FATAL_ERROR
+    "Unable to find toolchain in ${TOOLCHAIN_HOME} "
+    "Run `west espressif install` to download it. Then, export to path accordingly.")
```
@tejlmand This is current error if toolchain not found by either missing its installation or having previous installation version.

tejlmand replied:
yes, I see that.
But this just ask user to re-install, but a user already having the toolchain installed should not need to re-install it.

cfriedt commented:
Looks like this will likely be going into a dot release.

sylvioalves commented:
> Looks like this will likely be going into a dot release.

@cfriedt I am working on it. Hope there is still room for another round.

sylvioalves committed:
cmake: esp32: use zephyr defined python tool

This configures esptool, partition generator and
external bootloader build to use python tool defined
in Zephyr's context environment. This fixes error
regarding missing /usr/bin/python when python2.7 is not
installed in users's machine.

Partition table offset is also updated as part of
this change.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

sylvioalves commented:
@tejlmand 

> I think we should ensure that if users have a setup with the old style then that will still be supported.  

This update allows using already configured environment as you requested.

> But this just ask user to re-install, but a user already having the toolchain installed should not need to re-install it.  

I kept this message and it will only be displayed when no toolchain is exported into path. 

> because I only have python3 and no python link.
> Why not use the python that was already invoked with west ?  

I also added into this PR the necessary changes to use `${PYTHON_EXECUTABLE}` as part of the scripts. There were a few places where `python` was hardcoded.  

@cfriedt However, I had to update `west.yml` in order to get things in place in hal side. Now, #38543 needs to be merged before this MR. Once that gets merged, we can merge this. Do you think it can be accomplished before 2.7.0?

cfriedt commented:
The release team says they would prefer it to go into a dot release.

sylvioalves commented:
> The release team says they would prefer it to go into a dot release.

Ok, so we wait. Thanks.

tejlmand reviewed (changes requested):
Thanks for the update, I think I spotted some flaws.

sylvioalves commented on cmake/toolchain/espressif/generic.cmake:
```diff
@@ -10,23 +10,47 @@ set(COMPILER gcc)
+
+  if(ESPRESSIF_DEPRECATED_PATH)
+    set(CROSS_COMPILE ${ESPRESSIF_TOOLCHAIN_PATH}/bin/${CROSS_COMPILE_TARGET}-)
+    set(SYSROOT_DIR   ${ESPRESSIF_TOOLCHAIN_PATH}/${SYSROOT_TARGET}/${SYSROOT_TARGET})
```
Fixed!

tejlmand reviewed (approved):
lgtm

cfriedt commented:
Added the `DNM` label just until we are out of the stabilization period

sylvioalves commented:
@cfriedt Only rebased to fix merge conflict.
